src/System.Memory/src/System/Buffers/Text/Utf8Parser/Utf8Parser.Integer.Unsigned.D.cs

```diff
@@ -9,225 +9,276 @@ public static partial class Utf8Parser
+                    do
                     {
-                        bytesConsumed = index;
-                        value = (ushort)(parsedValue);
-                        return true;
-                    }
-                    parsedValue = parsedValue * 10 + nextDigit;
-                }
-                for (int index = ParserHelpers.Int16OverflowLength - 1; index < text.Length; index++)
-                {
-                    uint nextDigit = text[index] - 48u; // '0'
-                    if (nextDigit > 9)
-                    {
-                        bytesConsumed = index;
-                        value = (ushort)(parsedValue);
-                        return true;
-                    }
-                    // If parsedValue > (ushort.MaxValue / 10), any more appended digits will cause overflow.
-                    // if parsedValue == (ushort.MaxValue / 10), any nextDigit greater than 5 implies overflow.
-                    if (parsedValue > ushort.MaxValue / 10 || (parsedValue == ushort.MaxValue / 10 && nextDigit > 5))
-                    {
-                        bytesConsumed = 0;
-                        value = default;
-                        return false;
-                    }
-                    parsedValue = parsedValue * 10 + nextDigit;
+                        index++;
+                        if ((uint)index >= (uint)text.Length)
+                            goto Done;
+                        digit = text[index];
+                    } while (digit == '0');
+
+                    digit = digit - 48u; // '0'
+                    if (digit > 9)
+                        goto Done;
                 }
+
+                answer = digit;
+                index++;
+
+                if ((uint)index >= (uint)text.Length)
+                    goto Done;
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
+                    goto Done;
+                index++;
+                answer = 10 * answer + digit;
+
+                if ((uint)index >= (uint)text.Length)
+                    goto Done;
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
+                    goto Done;
+                index++;
+                answer = 10 * answer + digit;
+
+                if ((uint)index >= (uint)text.Length)
+                    goto Done;
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
+                    goto Done;
+                index++;
+                answer = 10 * answer + digit;
+
+                // Potential overflow
+                if ((uint)index >= (uint)text.Length)
+                    goto Done;
+                digit = text[index] - 48u; // '0';
+                if (digit > 9)
+                    goto Done;
+                answer = 10 * answer + digit;
+                if (answer > ushort.MaxValue)
+                    goto FalseExit; // Overflow
+                index++;
```

Comment:
same nit here and elsewhere. keep the two lines together everywhere:
```C#
index++;
answer = 10 * answer + digit;
```

Reply:
Moved.